File: Jack mackerel/Age2Length.cpp
//-----TRANSFORMATION FUNCION AGE->LENGTH--------------------------------------------------
FUNCTION Get_Age2length
// This subroutine allows convert an age composition to length composition. For example: if there is a matrix C(1,nyears,1,nages),
// the vectorial operation: Cl=C*Prob_length,  returns a matrix Cl(1,nyears,1,nlength) whose sum over all the lengths is the same
// the sum over all age groups..
// (Cristian Canales)
// by default values
// Linf=Linfprior;// Asymptotic length
// k_coeff=kprior;
// Lo=Loprior;// first length (corresponds to first age-group)
// sdage=sdageprior;// coefficient of variation of length-at-age
// if some of these are estimated.
Linf    = mfexp(log_Linf);
k_coeff = mfexp(log_k);
Lo      = mfexp(log_Lo);
sdage   = mfexp(log_sdage);
for (r=1;r<=ngrowth;r++)
{
    int i, j;
    mu_age(r,1)=Lo(r); // first length (modal)
    for (i=2;i<=nages;i++)
        mu_age(r,i) = Linf(r)*(1.-exp(-k_coeff(r)))+exp(-k_coeff(r))*mu_age(r,i-1); // the mean length by age group
    wt_age_vb(r) = lw_a * pow(mu_age(r), lw_b);
    maturity_vb(r) = 1/(1 + exp(32.93 - 1.45*mu_age(r)));
    sigma_age(r)=sdage(r)*mu_age(r); // standard deviation of length-at-age
    P_age2len(r) = ALK( mu_age(r), sigma_age(r), len_bins);
}
FUNCTION dvar_matrix ALK(dvar_vector& mu, dvar_vector& sig, dvector& x)
    //RETURN_ARRAYS_INCREMENT();
    int i, j;
dvariable z1;
dvariable z2;
int si,ni; si=mu.indexmin(); ni=mu.indexmax();
int sj,nj; sj=x.indexmin(); nj=x.indexmax();
dvar_matrix pdf(si,ni,sj,nj);
double xs;
pdf.initialize();
for(i=si;i<=ni;i++) //loop over ages
{
    for(j=sj;j<=nj;j++) //loop over length bins
    {
        if (j<nj)
            xs=0.5*(x[sj+1]-x[sj]);  // accounts for variable bin-widths...?
        z1=((x(j)-xs)-mu(i))/sig(i);
        z2=((x(j)+xs)-mu(i))/sig(i);
        pdf(i,j)=cumd_norm(z2)-cumd_norm(z1);
    }//end nbins
    pdf(i)/=sum(pdf(i));
}//end nage
//RETURN_ARRAYS_DECREMENT();
return(pdf);

File: Jack mackerel/Age2Length.R
# Age and length bins
nages = 12
len_bins = c(10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50)
nlengths = length(len_bins)
mu_age <- c()

# Growth parameters
Linf    = 73.56
k_coeff = 0.16
Lo      = 13.56
sdage   = 0.09
lw_a = 0.007778994e-3 # LW parameters from Talcahuano sampling, 2008
lw_b = 3.089248476 # LW parameters from Talcahuano sampling, 2008

# Calculate mean length-at-age
mu_age[1]=Lo # first length (modal)
for (i in 2:nages){
    mu_age[i] = Linf*(1.-exp(-k_coeff))+exp(-k_coeff)*mu_age[i-1] # the mean length by age group
}
wt_age_vb = lw_a * mu_age^lw_b
maturity_vb = 1/(1 + exp(32.93 - 1.45*mu_age))
sigma_age=sdage*mu_age # standard deviation of length-at-age
P_age2len = ALK(mu_age, sigma_age, len_bins)


# Calculate transition matrix
pdf <- matrix(0, nages, nlengths)
for(i in 1:nages){
    for(j in 1:nlengths){
        if(j<nlengths){
            xs=0.5*(len_bins[2]-len_bins[1])  # accounts for variable bin-widths...?
        }
        z1=((len_bins[j]-xs)-mu_age[i])/sigma_age[i]
        z2=((len_bins[j]+xs)-mu_age[i])/sigma_age[i]
        pdf[i,j]=pnorm(z2)-pnorm(z1)
    }
    pdf[i,] = pdf[i,]/sum(pdf[i,])
}

write.csv(pdf, file = "JJM_pdf.csv")
